pkgs/tools/compression/gzip/default.nix

```diff
@@ -17,6 +17,27 @@ stdenv.mkDerivation rec {
 
   makeFlags = [ "SHELL=/bin/sh" "GREP=grep" ];
 
+  # Many gzip executables are shell scripts that depend upon other gzip
+  # executables being in $PATH.  Rather than try to re-write all the
+  # internal cross-references, just add $out/bin to PATH at the top of
+  # all the executables that are shell scripts.
+  preFixup = ''
+    one_line() { IFS= read -r line && printf "%s\n" "$line"; }
+    remaining_lines() { while one_line; do :; done; }
+    for f in $out/bin/*; do
+      executable_content=$(< "$f")
+      if [[ "''${executable_content:0:10}" == $'#!/bin/sh\n' ]];then
+        {
+          one_line
+          echo "PATH=\"$out/bin:\$PATH\""
+          remaining_lines
+        } < "$f" > "$f.new"
+        mv "$f.new" "$f"
+        chmod +x "$f"
+      fi
+    done
```

Comment:
Any reason not to use wrapProgram?

Reply:
I was trying to minimise dependencies, since gzip seems like the sort of thing that other low-level stuff would want to depend upon.  Bash is already in use here, and this can be done in pure bash, so I did it in pure bash.

At your suggestion [I tried it with wrapProgram here](https://github.com/chkno/nixpkgs/commit/8456e0854d0e21a2418f06f1d61351edc34a2002), and it didn't immediately blow up, so maybe that's fine?

I don't know which is more valuable here: minimal dependencies or the simpler build script.

Reply:
What dependencies does `wrapProgram` bring in? From what I can see it's just a shell?

Reply:
```
$ nix-store -q --references "$(nix-instantiate --expr 'with import <nixpkgs> {}; makeWrapper')"
/nix/store/0s9...-make-wrapper.sh
/nix/store/9kr...-default-builder.sh
/nix/store/hla...-bash-4.4-p23.drv
/nix/store/jxg...-stdenv-linux.drv
/nix/store/l64...-hook.drv
```
But then `stdenv-linux` depends upon gzip:
```
$ nix-store -q --requisites "$(nix-instantiate --expr 'with import <nixpkgs> {}; makeWrapper')" | grep gzip
/nix/store/310...-gzip-1.10.tar.xz.drv
/nix/store/lqm...-gzip-1.10.drv
```

Reply:
Argh, well, I say keep this as-is then

Reply:
I think the set of dependencies that is always OK to use even in build scripts for low-level tools is the ones in `bootstrap-tools`. After all, everything requires tools to build, and Bash is no exception. :slightly_smiling_face: 

Those tools include, notably, GNU `sed` and the GNU coreutils:
```
$ ls $(nix-build -A stdenv.bootstrapTools)/bin
'['         cksum       env      install   nohup      rmdir    test
 ar         cmp         expand   join      nproc      sdiff    timeout
 as         comm        expr     kill      numfmt     sed      touch
 awk        coreutils   false    ld        objdump    seq      tr
 b2sum      cp          fgrep    link      od         sh       true
 base32     cpp         find     ln        paste      sleep    truncate
 base64     csplit      fmt      logname   patch      sort     tsort
 basename   cut         fold     ls        patchelf   split    tty
 basenc     date        g++      make      pr         stat     uname
 bash       dd          gawk     md5sum    printenv   stdbuf   unexpand
 bunzip2    df          gcc      mkdir     printf     strip    uniq
 bzip2      diff        grep     mkfifo    ptx        stty     unlink
 cat        diff3       groups   mknod     pwd        sum      uptime
 chcon      dircolors   gunzip   mktemp    ranlib     sync     wc
 chgrp      dirname     gzip     mv        readelf    tac      xargs
 chmod      du          head     nice      readlink   tail     xz
 chown      echo        hostid   nl        realpath   tar      yes
 chroot     egrep       id       nm        rm         tee
```

So I think that opens up some ways to keep the script simpler. For example something like (untested):
```nix
  preFixup = ''
    for f in $out/bin/*; do
      if head -1 "$f" | grep -qxF '#!/bin/sh'; then
        sed -i '2iPATH='$out'/bin:"$PATH"' "$f"
      fi
    done
  ''
```

That should do exactly the same thing as the all-Bash implementation in the current version of the PR.

Reply:
Cool, thanks!  I shortened it to a single `sed -i`.